staging/proj.finassistant/manifest.tt

```diff
@@ -0,0 +1,139 @@
+  #_[formatted=[{type="text",text="The PE ratio of ${company} is ${pe}."}]]
+  #[doc="Get PE ratio"];
+
+  query ps(in req company: Entity(tt:stock_id)
+    		   #_[prompt="What company do you want to check?"],
+               out ps: Number
+               )
+  #_[canonical="P/S ratio"]
+  #_[confirmation="P/S ratio of ${company}"]
+  #_[formatted=[{type="text",text="The P/S ratio of ${company} is ${ps}."}]]
+  #[doc="Get P/S ratio"];
+
+  query revenue(in req company: Entity(tt:stock_id)
+    		   #_[prompt="What company do you want to check?"],
+               out revenue: Currency
+               )
+  #_[canonical="revenue"]
+  #_[confirmation="revenue of ${company}"]
+  #_[formatted=[{type="text",text="The revenue of ${company} is ${revenue}."}]]
+  #[doc="Get revenue"];
+
+  query earnings(in req company: Entity(tt:stock_id)
+             #_[prompt="What company do you want to check?"],
+               out earnings: Number
+               )
+  #_[canonical="earnings"]
+  #_[confirmation="earnings of ${company}"]
+  #_[formatted=[{type="text",text="The earnings of ${company} is ${earnings}."}]]
+  #[doc="Get earnings"];
+
+  query market_cap(in req company: Entity(tt:stock_id)
+    		   #_[prompt="What company do you want to check?"],
+               out market_cap : Currency
+               )
+  #_[canonical="market cap"]
+  #_[confirmation="Market cap of ${company}"]
+  #_[formatted=[{type="text",text="The market cap of ${company} is ${market_cap}."}]]
+  #[doc="Get market cap"];
+
+  query ebitda(in req company: Entity(tt:stock_id)
+    		   #_[prompt="What company do you want to check?"],
+               out ebitda : Currency
+               )
+  #_[canonical="ebitda"]
+  #_[confirmation="ebitda of ${company}"]
+  #_[formatted=[{type="text",text="The ebitda of ${company} is ${ebitda}."}]]
+  #[doc="Get ebitda"];
+
+  query operating_profit(in req company: Entity(tt:stock_id)
+    		   #_[prompt="What company do you want to check?"],
+               out operating_profit : Currency
+               )
+  #_[canonical="operating_profit"]
+  #_[confirmation="operating_profit of ${company}"]
+  #_[formatted=[{type="text",text="The operating_profit of ${company} is ${operating_profit}."}]]
+  #[doc="Get operating_profit"];
+
+  query gross_profit(in req company: Entity(tt:stock_id)
+    		   #_[prompt="What company do you want to check?"],
+               out gross_profit : Currency
+               )
+  #_[canonical="gross_profit"]
+  #_[confirmation="gross_profit of ${company}"]
+  #_[formatted=[{type="text",text="The gross_profit of ${company} is ${gross_profit}."}]]
+  #[doc="Get gross_profit"];
+
+  query create_report(in req data: Object
+    		   #_[prompt="What do you want to name your report?"]
+               )
+  #_[canonical="Create report"]
+  #_[confirmation="Created report ${name}"]
+  #[doc="create report"];
+
+  query display_report(in req name: String
+    		   #_[prompt="What report do you want?"]
+               )
```

Comment:
This function needs to return "something".

I am not sure exactly how to implement this feature. One way is to return a link to the report that your frontend knows how to display. Another way is to return a ThingTalk program with the report, so you reexecute the ThingTalk program in your frontend and get the updated data.
If you go with the link/thingtalk program, you'd write something like:
```
entity report;
list query report(out id : Entity(proj.financeassistant:report),
                          out created_on : Date,
                          out companies : Array(Entity(tt:stock_id)),
                          out fields : Array(Enum(revenue,earnings,market_cap, ...))
                          out link : Entity(tt:url))
```
Note that everything is an output. So you can pull up the report by name with:
```
@proj.financeassistant.report(), id =~ "my report";
```
(`id` is of type Entity so it has both an identifier and a name, `=~` is the soft-match operator and looks at the name)
but you can also pull up all reports of a given company:
```
@proj.financeassistant.report(), contains(companies, "goog"^^tt:stock_id);
```
or all market_cap reports:
```
@proj.financeassistant.report(), contains(fields, enum market_cap);
```
These are all relatively easy to express in natural language, and if you write the query this way and write enough annotations on each output parameter, you don't have to do anything for NL, Genie takes care of making up all the possible combinations. Also implementation wise it's quite easy: you can just dump the whole list of reports out of the Thingpedia function, and let the ThingTalk runtime do the actual filtering.